### PR TITLE
ethereum-executor: port evmone precompiles into the module (split 1/4)

### PR DESCRIPTION
**Pull request by morebtcg**
Part of splitting the bcos-evm-independent refactor (原 #5402) into small PRs, each within the 666-line `check-commit.sh` limit.

**本 PR（1/4）：移植 evmone 预编译层（纯新增，零行为变化）**
- 新增 `ethereum-executor/EVMPrecompiles.h/.cpp`：将 `bcos-evm/eth/state/precompiles.cpp` 移植进 `bcos::executor_v1::eth::evm` 命名空间（带 SYNC CONTRACT 头注释，与上游保持同步）。
- `ethereum-executor/CMakeLists.txt`：把 `EVMPrecompiles.cpp` 加入库并链接 `evmone::precompiles`。
- 未接线：旧 `BCOS2Evmone`/`StorageStateView` 桥保持不变，本 PR 不改变任何行为。

valid insertions ≈ 573（<= 666）。本 PR 与 #5412 可并行合入；后续 pr-3（EthereumHost）依赖本 PR 与 #5412。

### CODE REVIEW

File: ethereum-executor/EVMPrecompiles.cpp [hunk 1]
```diff
@@ -0,0 +1,847 @@
+/// @file EVMPrecompiles.cpp
```
Comment (kyonRay) [inline ethereum-executor/EVMPrecompiles.cpp:5]: **Must-fix: the evmone copyright and SPDX notice was dropped from a verbatim derivative.**

Upstream `test/state/precompiles.cpp` opens with:

```
// evmone: Fast Ethereum Virtual Machine implementation
// Copyright 2022 The evmone Authors.
// SPDX-License-Identifier: Apache-2.0
```

`precompiles.hpp` carries the same, and `precompiles_internal.hpp` carries the 2024 variant. This file and `EVMPrecompiles.h` carry none of them — `grep -i 'spdx\|copyright\|apache\|licen'` over both new files returns nothing.

Apache-2.0 section 4(c) requires retaining all copyright, patent, trademark and attribution notices from the source form in any derivative work you distribute, and this file is about as derivative as it gets: the byte comparison I ran shows the only changes are the namespace, the two removed build-macro branches, one deleted TODO and two explicit aggregate initializers. The repo's own precedent agrees — `bcos-evm/bcos-evm/eth/state/precompiles.cpp` is the same upstream file and it kept the header intact (it is byte-identical to upstream, header included).

Fix: put the three upstream lines back at the top of both `EVMPrecompiles.cpp` and `EVMPrecompiles.h`, above the `/// @file` block. That is the whole change. Same for the 2024 notice if you want to be precise about the `_internal.hpp` provenance, since this header merges both upstream headers.

---

**Separately, on the SYNC CONTRACT text itself.**

The note says the intended differences are "the namespace, the removed libsecp256k1/GMP build-time alternatives, the `namespace crypto` alias and the local to_underlying helper". There is one more, at the bottom of `call_precompile`:

```cpp
const evmc_result result{status_code, status_code == EVMC_SUCCESS ? gas_left : 0, 0,
    output_data, output_size,
    [](const evmc_result* res) noexcept { delete[] res->output_data; },
    evmc_address{},  // create_address
    {}               // padding
};
```

Upstream stops after the release callback. The two added initializers are a no-op — aggregate initialization value-initializes the remaining members, which is exactly `evmc_address{}` and `{}` — so I assume this is here to silence `-Wmissing-field-initializers` under `-Werror`. That is a fine reason; the problem is only that the contract advertises itself as exhaustive. A future maintainer diffing against upstream will hit this line, find it undocumented, and have to re-derive that it is harmless. One clause in the note fixes that.

(The removed TODO comment is not worth listing.)

File: ethereum-executor/CMakeLists.txt [hunk 2]
```diff
@@ -8,13 +8,14 @@ add_library(ethereum-executor
...
     StorageStateView.cpp
+    EVMPrecompiles.cpp
 )
 target_include_directories(ethereum-executor PUBLIC
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
     $<INSTALL_INTERFACE:include/ethereum-executor>)
 target_link_libraries(ethereum-executor PUBLIC
     bcos-evm-eth bcos-framework bcos-protocol
-    evmone::evmone intx::intx TBB::tbb)
+    evmone::evmone evmone::precompiles intx::intx TBB::tbb)
```
Comment (kyonRay) [inline ethereum-executor/CMakeLists.txt:18]: Two notes on this target, neither blocking.

**`EVMPrecompiles.cpp` has no caller yet.** Nothing in the tree references `EVMPrecompiles.h`, `eth::evm::call_precompile` or `eth::evm::is_precompile`. Because the file is in `add_library` and the target sets `UNITY_BUILD OFF`, CI will compile it standalone and catch syntax or link errors — but no test exercises it, so a behavioural difference from `bcos-evm`'s copy would pass green. That is inherent to a 1/4 split and I am not asking for anything here; it is just the right expectation to set for what this PR's CI result proves.

**`evmone::precompiles` is already implied.** `evmoneConfig.cmake:24` lists it in `evmone::evmone`'s `INTERFACE_LINK_LIBRARIES`, and `bcos-evm/CMakeLists.txt:91` links it explicitly too — so `ethereum-executor` was already getting it transitively through `bcos-evm-eth`. Naming it here is still the right call now that this target uses the `evmone_precompiles/*` primitives directly: it stops the dependency from disappearing the moment split 3/4 drops `bcos-evm-eth` from this line. No change wanted, just recording that I checked the target exists and is exported.

**Review by kyonRay — CHANGES_REQUESTED** [review]
## Review — split 1/4 (`ethereum-executor` precompiles port)

**Verdict: CHANGES_REQUESTED — one must-fix, and it is a three-line one.** Everything technical about this port checks out; the blocker is a licensing-attribution omission, not a defect in the code.

The split itself did exactly what it was supposed to. Recomputing the repo's own gate on this head: `valid_insertions = 573` against `insert_limit = 666` (968 raw insertions, minus 2 new files x 20, minus 59 comment / 153 blank / 124 brace-only / 19 include lines). The monolithic #5402 was 809. Base `1a62e3d92`, 0 commits behind, single commit.

### What I verified, and how

I did a three-way byte comparison rather than reading the port for plausibility:

| pair | result |
|---|---|
| `evmone v0.21.0 test/state/precompiles.{cpp,hpp}` + `precompiles_internal.hpp` vs `bcos-evm/bcos-evm/eth/state/*` | **byte-identical** (the existing bcos-evm copy is a verbatim vendor of upstream) |
| upstream `precompiles.cpp` vs this PR's `EVMPrecompiles.cpp`, after normalizing `stdx::to_underlying` -> `to_underlying` | **only the differences listed below** |
| this PR's files vs the same files at #5402 head `b1023cca0` | **byte-identical** — the split carries the already-reviewed content unchanged |

The complete set of differences from upstream is:

1. `#ifdef EVMONE_PRECOMPILES_LIBSECP256K1` and `#ifdef EVMONE_PRECOMPILES_GMP` branches removed, keeping the evmone implementations. **Verified dead in this build**: neither macro is defined anywhere in the repo outside the `#ifdef`s in `bcos-evm/bcos-evm/eth/state/precompiles.cpp` itself, and the evmone vcpkg port does not define them either. So `ecrecover` and `expmod` resolve to the same implementation in both copies — no behavioural divergence between executor v1 and v2 here.
2. `namespace evmone::state` -> `bcos::executor_v1::eth::evm`, plus the local `to_underlying` and the `namespace crypto = evmone::crypto` alias.
3. One dropped `// TODO:` comment about `evmc::Result(std::unique_ptr)`.
4. The `evmc_result` aggregate gained two explicit initializers (`create_address`, `padding`). Semantically a no-op — upstream's six-initializer aggregate value-initializes exactly those two — but see the inline note: the SYNC CONTRACT claims to enumerate every intended difference and this one is missing from it.

Every gas-schedule function (`*_analyze`) and every execute function is otherwise character-for-character upstream, including the Osaka-era `expmod` repricing, the EIP-7823 length limit and the BLS12 sizing constants. `is_precompile`'s availability lookup table is identical modulo the `stdx::` qualifier. All eleven anonymous-namespace entities are referenced, so nothing trips `-Wunused-function` / `-Wunused-const-variable` under `-Werror`.

### Scope facts that re-scale everything above

- **No caller.** Nothing in the tree references `EVMPrecompiles.h`, `eth::evm::call_precompile` or `eth::evm::is_precompile`. The `.cpp` *is* compiled (it is in `add_library`, and the target has `UNITY_BUILD OFF`), so CI will catch a compile error — but it will not catch a behaviour change, because no test reaches it. The port is inert until split 3/4 wires it in.
- **`bcos-evm-eth` is still on the link line**, so after this merges the binary carries two copies of the precompile dispatch. They are in different namespaces, so there is no ODR collision — that part of the header comment is accurate and I checked it. This is the "copy, not share" point from #5402 round 1; the SYNC CONTRACT note is the mitigation we agreed on, and I am not re-litigating it here. Worth saying out loud though: a comment is a convention, not a guard. If these two ever have to coexist for more than a release, a test that runs the same vector set through both entry points would turn the convention into something CI can enforce.
- `evmone::precompiles` on the link line is redundant — `evmoneConfig.cmake` already lists it in `evmone::evmone`'s `INTERFACE_LINK_LIBRARIES`, and `bcos-evm/CMakeLists.txt:91` links it too. Redundant-but-explicit is the right call for a target that now uses those primitives directly; no change wanted.

### Checked, not a problem

- **Gas schedules**: no drift in any `*_analyze` function. This was the thing worth checking most and it is clean.
- **`max_blob_gas_per_block`-style promotion bugs**: not in this file.
- `-Wmissing-declarations` on `ecrecover_execute_evmone` / `expmod_execute_evmone`, which lost their header declarations when the two upstream headers were merged into one: that warning is not in the repo's set (`-Wall -Wextra -pedantic -Werror`), and both have external linkage in a distinct namespace, so nothing collides.
- **clang-format**: zero lines over the 100-column limit in either new file. (My local clang-format is 22.1.8 and produces some include-ordering churn; that is version skew, not a violation, so I am not reporting it.)

### Verification boundary

Everything above is read from the committed blobs at `395dfb0c2` and diffed against `vcpkg/buildtrees/evmone/src/v0.21.0-f20dfcaeef.clean/test/state/` and `bcos-evm/bcos-evm/eth/state/` on disk. I did not build this branch and did not run any test. The claim "the port is behaviourally identical" rests on the byte comparison plus the verified absence of the two removed build macros — not on execution.

File: ethereum-executor/EVMPrecompiles.cpp (hunk 1 above)
Comment (morebtcg) [reply ethereum-executor/EVMPrecompiles.cpp:5]: Fixed in 2ed020a56. Added the upstream evmone notice (Copyright 2022 The evmone Authors. / SPDX-License-Identifier: Apache-2.0) to the top of both `EVMPrecompiles.cpp` and `EVMPrecompiles.h` (the .h additionally credits the 2024 `precompiles_internal.hpp` portions), and updated the SYNC CONTRACT note to list the two explicit aggregate initializers on the `evmc_result` in `call_precompile` (`evmc_address{}` for `create_address` and `{}` for `padding`) as an intended difference — they value-initialize the members this evmc version requires under `-Werror=missing-field-initializers`.

File: ethereum-executor/CMakeLists.txt (hunk 2 above)
Comment (morebtcg) [reply ethereum-executor/CMakeLists.txt:18]: No code change needed. The file has no callers yet only because this PR is the first of a 4-PR split of the ethereum-executor refactor; `EVMPrecompiles.cpp` is added as a source in this PR and the executor starts using it in a later PR. Naming `evmone::precompiles` explicitly is intentional so the dependency is declared rather than only transitive.

File: ethereum-executor/EVMPrecompiles.h [hunk 3]
```diff
@@ -0,0 +1,125 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2022 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+// Portions from evmone test/state/precompiles_internal.hpp: Copyright 2024 The
+// evmone Authors.
+
+/// @file EVMPrecompiles.h
+/// @brief Ported evmone test/state precompiles (precompiles.hpp +
+///        precompiles_internal.hpp), renamed into the
+///        `bcos::executor_v1::eth::evm` namespace so the ethereum-executor no
```
Comment (ywy2090) [inline ethereum-executor/EVMPrecompiles.h:10]: ⚠️ [IMPORTANT] [OPTIONAL] "no longer depends on the bcos-evm library" claim is ahead of implementation

**Impact**: this split's CMakeLists.txt:17 still links bcos-evm-eth (containing the live `evmone::state::call_precompile`, host.cpp:366-367), so this is a parallel port, not a decoupling. A reader taking this comment at face value would misjudge the remaining splits.
**Evidence**: `git show 49c9f925:ethereum-executor/CMakeLists.txt:17`.
**Suggestion**: reword to "will eventually":
```suggestion
/// @brief Ported evmone test/state precompiles (precompiles.hpp +
///        precompiles_internal.hpp), renamed into the
///        `bcos::executor_v1::eth::evm` namespace so that, once the
///        remaining splits wire this module in, ethereum-executor will no
///        longer depend on the bcos-evm library (and never ODR-collides
///        with it in a shared binary).
```

File: ethereum-executor/EVMPrecompiles.cpp [hunk 4]
```diff
@@ -0,0 +1,854 @@
...
+    const auto [analyze, execute] = EXECUTION_LOOKUP_TABLE[to_lookup_index(msg.code_address)];
+
+    const bytes_view input{msg.input_data, msg.input_size};
+    const auto [gas_cost, max_output_size] = analyze(input, rev);
+    const auto gas_left = msg.gas - gas_cost;
+    if (gas_left < 0)
+        return evmc::Result{EVMC_OUT_OF_GAS};
+
+    // Allocate buffer for the precompile's output and pass its ownership to evmc::Result.
+    const auto output_data = new (std::nothrow) uint8_t[max_output_size];  // TODO: handle nullptr.
```
Comment (ywy2090) [inline ethereum-executor/EVMPrecompiles.cpp:843]: 💡 [SUGGESTION] [OPTIONAL] call_precompile allocation-failure nullptr unguarded

**Impact**: `new (std::nothrow) uint8_t[max_output_size]` returning nullptr leads to writing into a null pointer (memory pressure / large input). Inherited upstream TODO, not a port regression — cheapest to fix now (no caller).
**Suggestion**:
```suggestion
    const auto output_data = new (std::nothrow) uint8_t[max_output_size];
    if (output_data == nullptr)
    {
        return evmc::Result{EVMC_OUT_OF_GAS};
    }
```

**Review by ywy2090 — APPROVED** [review]
Round 2 · PR #5411, at `49c9f925`.

APPROVE — 0 critical / 0 important blocking. The kyonRay must-fix (evmone copyright header) is fixed; line-by-line comparison against the bcos-evm copy shows zero unintended drift. 3 IMPORTANT findings are inherent to the split / ahead-of-implementation claims — recommended fixes, not merge blockers.

CI: check commit ✓, Hound ✓, builds/coverage/L2 pending; Codacy ✗ (quality service, unrelated to functionality).

## Previous findings

| # | severity | finding | source | status |
|---|----------|---------|--------|--------|
| A | must-fix | evmone copyright/SPDX dropped | kyonRay prev round | **fixed** — both head files carry Copyright 2022/2024 + SPDX Apache-2.0 |
| B | note | EVMPrecompiles.cpp has no caller | kyonRay prev round | known (inherent to split 1/4) |

### Finding A · IMPORTANT — "no longer depends on bcos-evm" claim is ahead of implementation
**Impact**: EVMPrecompiles.h:8-11 claims the ethereum-executor no longer depends on bcos-evm, but CMakeLists.txt:17 still links bcos-evm-eth — a STATIC lib containing the live `evmone::state::call_precompile` (host.cpp:366-367). Split 1/4 is a parallel port, not a decoupling: two full precompile dispatch layers (18 analyze + 18 execute + gas tables) will coexist in the production binary (libinitializer pulls ethereum-executor into init). Namespace isolation avoids ODR collision but not code duplication.
**Evidence**: `git show 49c9f925:ethereum-executor/CMakeLists.txt:17`; `EVMPrecompiles.h:8-11`.
**Suggestion**: Reword the header/design comment to "will eventually no longer depend"; drop bcos-evm-eth from the link list in split 2/4 (when the first caller lands).

### Finding B · IMPORTANT — zero callers, zero tests; behavioral fidelity has no anchor
**Impact**: EVMPrecompiles.h is included by nothing in the repo (only its own .cpp), call_precompile/is_precompile have no caller, PR adds no test. All green tests exercise the bcos-evm old path (host.cpp → evmone::state); any gas-constant/boundary bug in the port passes CI silently. Fidelity rests on the SYNC CONTRACT comment (prose only).
**Evidence**: `git grep -l EVMPrecompiles.h pr-5411-head` → only .cpp/.h; ethereum-executor/tests builds only eest-runner, never invokes the new symbols.
**Suggestion**: Write a differential parity test in THIS split — bcos-evm-eth is still linked, so `bcos::executor_v1::eth::evm::call_precompile == evmone::state::call_precompile` (same status/gas/output) can be asserted now. Once a later split removes the dependency, this parity test can never be written again.

### Finding C · IMPORTANT — duplicated dispatch layer maintenance debt (perf agent)
**Impact**: Two complete dispatch + gas-table layers coexist in one binary; a fork gas repricing or new precompile edited in one silently diverges from the other with no diagnostic. Gas constants are fork-sensitive correctness data.
**Evidence**: bcos-evm precompiles.cpp compiles into bcos-evm-eth (live code), the new module copies it verbatim; heavy crypto kernels are shared (both link evmone::precompiles), only the thin dispatch layer is duplicated.
**Suggestion**: PR description should state why the dispatch is not shared (decoupling is the long-term goal, temporary cost); add a drift-guard test (see B) so a repricing fails CI.

### Finding D · SUGGESTION — call_precompile heap alloc + nullptr unhandled
**Impact**: Every precompile call does `new (std::nothrow)` for the output buffer (hot-path malloc/free), and `// TODO: handle nullptr` does not guard allocation failure. Inherited from upstream design, not a port regression — but cheapest to fix now (no caller).
**Suggestion**: When the executor wires in, use the analyze→execute two-phase API with a caller-owned scratch buffer (no signature change); add the nullptr check.

### Finding E · NIT — misc
- Include self-sufficiency half-done: explicitly adds <array>/<bit>/<cassert>/<limits>/<span>, misses <algorithm>/<cstring>/<ranges>/<vector>/<optional> (works via evmone/intx transitives)
- ecrecover_execute_evmone / expmod_execute_evmone have external linkage but aren't declared in EVMPrecompiles.h (bcos-evm internal declares them)
- Two evmc headers coexist: vcpkg standalone evmc_arm64-osx lacks EVMC_OSAKA; compilation implicitly depends on evmone's own evmc include order (environment-level fragile point; suggest pinning in vcpkg.json)
- PrecompileId enum declared but unused in this module (upstream also only declares)

## Verification boundary
- Done: 4-dimension parallel agents, all findings verified against `git show pr-5411-head:`; correctness agent line-by-line diff vs bcos-evm (only the 8 declared differences) + measured all 18 precompile gas costs / expmod bounds / is_precompile address domain / call_precompile failure paths; architecture agent confirmed namespace ODR-safety and acyclic module layering; perf agent confirmed heavy crypto is shared, no per-call extra copies; test agent verified the SYNC CONTRACT claim is accurate and confirmed the zero-caller gap
- Not done: no compile/test run (builds pending); differential parity test not written (recommended for this split); Codacy fail specifics not fetched (network)
- Head check: LIVE == LOCAL == 49c9f925, passed before output

## Closing
Fix A (reword comment to "will eventually"), B (differential parity test — the only window), D (nullptr guard). C documents the duplicated-dispatch debt; E optional.

File: ethereum-executor/EVMPrecompiles.cpp [hunk 5]
```diff
@@ -0,0 +1,854 @@
...
+        };
+        std::array<Entry, LOOKUP_TABLE_SIZE> table{};
+        for (const auto& trait : traits)
+            table[trait.address] = {trait.analyze, trait.execute};
+        return table;
+    }();
+
+    assert(msg.gas >= 0);
+
+    const auto [analyze, execute] = EXECUTION_LOOKUP_TABLE[to_lookup_index(msg.code_address)];
```
Comment (Shichen-Wu) [inline ethereum-executor/EVMPrecompiles.cpp:834]: **🟡 观察 — `call_precompile` 无 `is_precompile` 前置检查，空槽位会触发空指针调用**

`EXECUTION_LOOKUP_TABLE[to_lookup_index(msg.code_address)]` 直接索引，未检查该地址是否真的是 precompile。`traits` 表不连续（0x01-0x11 之后跳到 0x0100），若 `msg.code_address` 低 16 位落在空槽位（如 0x0012、0x0099），会取到默认 `{nullptr, nullptr}`，随后 `analyze(input, rev)`（第 836 行）是空指针调用（UB）。

这是上游 evmone 的既有行为（移植保持 byte-identical），依赖调用方（后续 PR-3 的 EthereumHost）先做 `is_precompile` 检查。建议：① 在 `call_precompile` 开头加 `assert(is_precompile(rev, msg.code_address))`；② 或在 SYNC CONTRACT 注释中显式声明该调用契约，防止后续接线时遗漏检查。

**Review by Shichen-Wu — COMMENTED** [review]
## Review：PR #5411（split 1/4 — evmone precompiles 移植）

**纯移植 PR，质量高，无 P0/P1 阻塞问题**。我独立核对了 EVMPrecompiles.h（完整 18 个 precompile，`bcos::executor_v1::eth::evm` 命名空间避免 ODR 冲突）、EVMPrecompiles.cpp（与上游 byte-identical，`_evmone` 包装层断言 output_size）、evmone copyright/SPDX notice 已恢复（`49c9f92` 修复）、SYNC CONTRACT 注释清晰。kyonRay 已深度验证（upstream byte-identical），结论可靠。

补充 2 个未被覆盖的观察点（行内评论见下），均非阻塞。

File: ethereum-executor/EVMPrecompiles.cpp (hunk 4 above)
Comment (Shichen-Wu) [inline ethereum-executor/EVMPrecompiles.cpp:843]: **🟡 观察 — `new (std::nothrow)` 分配失败返回 nullptr 未处理**

`new (std::nothrow) uint8_t[max_output_size]` 若内存分配失败返回 nullptr，会直接传给 `execute`（第 846 行）导致崩溃。`max_output_size` 为 0 时 new[] 不失败，但大输出（如 BLS、modexp 的极大 modulus）配合内存压力时可能触发。上游同样有 `TODO: handle nullptr`——既然移植了，建议顺手处理：分配失败返回 `{EVMC_OUT_OF_MEMORY, 0}` 或至少 `assert(output_data != nullptr)`。

**Review by kyonRay — APPROVED** [review]
Round 2, at `49c9f925b`. **Approving — the must-fix is fully resolved and there is nothing new.**

The delta from `395dfb0c2` is exactly two hunks, both of them the things I asked for, and no code was touched. That matters: it means the three-way byte comparison from round 1 (upstream evmone v0.21.0 → `bcos-evm`'s vendored copy → this port) still holds unchanged, so the "behaviourally identical to upstream" conclusion carries over without re-deriving it.

### The attribution fix, checked against upstream

`EVMPrecompiles.cpp` now opens with the three lines from upstream `test/state/precompiles.cpp` — **character-for-character identical**. `EVMPrecompiles.h` carries the 2022 notice from `precompiles.hpp` plus a separate line crediting the 2024 `precompiles_internal.hpp` portions. That second line is more precise than what I asked for and it is the right call: this header genuinely merges two upstream headers with different notice years, and the repo's own precedent (`bcos-evm/bcos-evm/eth/state/precompiles.cpp`, also 2022, intact) only covers the simpler case.

Worth recording for anyone who wonders why this was a blocker: **there is no license-header check in CI**. `tools/.ci/check-commit.sh:18`'s `license_line=20` is an allowance in the insertion-count arithmetic, not a validator. So this was an Apache-2.0 §4(c) compliance matter end to end, and nothing automated would ever have caught it.

### The SYNC CONTRACT is exhaustive again

It now lists the two explicit aggregate initializers on the `evmc_result` in `call_precompile` (`evmc_address{}` for `create_address`, `{}` for `padding`) with the `-Werror=missing-field-initializers` reason. A future maintainer diffing this file against upstream will no longer hit an undocumented line and have to re-derive that it is a no-op.

### Mechanical checks at this head

- Repo gate recomputed: `valid_insertions = 573` against `insert_limit = 666`. Passes.
- Zero lines over the 100-column limit in either new file. The residual clang-format differences are include ordering and sub-limit rewrapping — the same version-skew category I noted last round, and `check-commit.sh:47-50` runs `clang-format -i` in place and only fails if clang-format itself exits non-zero, so CI does not gate on them either.
- Base `1a62e3d92`, 0 commits behind.

Everything from round 1 that I marked "checked, not a problem" — no caller yet, the two coexisting precompile dispatches separated by namespace, `evmone::precompiles` being redundant-but-explicit on the link line — is untouched by this delta and stands as written.

**Verification boundary:** read from the committed blob at `49c9f925b` and diffed against the upstream evmone trees on disk. I did not build this branch and ran no test.
